File: crates/ncode-python/python/ncode/query.py
"""Fluent, injection-safe SQL builder — the Python mirror of the Rust builder."""

from __future__ import annotations

from typing import Optional, Union

Param = Union[int, float, str, bool, None]


class Query:
    """Build a ``SELECT`` statement fluently.

    >>> sql, params = (
    ...     Query.table("users").select("id", "name").filter("age >= ?", 30).limit(10).build()
    ... )
    >>> sql
    'SELECT id, name FROM users WHERE age >= ? LIMIT 10'
    """

    def __init__(self, table: str):
        self._table = table
        self._cols: list[str] = []
        self._wheres: list[str] = []
        self._params: list[Param] = []
        self._order: Optional[str] = None
        self._limit: Optional[int] = None
        self._offset: Optional[int] = None

    @classmethod
    def table(cls, name: str) -> "Query":
        return cls(name)

    def select(self, *columns: str) -> "Query":
        self._cols = list(columns)
        return self

    def filter(self, predicate: str, *params: Param) -> "Query":
        self._wheres.append(predicate)
        self._params.extend(params)
        return self

    def order(self, column: str, direction: str = "ASC") -> "Query":
        self._order = f"{column} {direction}"
        return self

    def limit(self, n: int) -> "Query":
        self._limit = n
        return self

    def offset(self, n: int) -> "Query":
        self._offset = n
        return self

    def build(self) -> tuple[str, list[Param]]:
        cols = ", ".join(self._cols) if self._cols else "*"
        sql = f"SELECT {cols} FROM {self._table}"
        if self._wheres:
            sql += " WHERE " + " AND ".join(self._wheres)
        if self._order:
            sql += f" ORDER BY {self._order}"
        if self._limit is not None:
            sql += f" LIMIT {self._limit}"
        if self._offset is not None:
            sql += f" OFFSET {self._offset}"
        return sql, list(self._params)

File: src/crates/ncode-node/ts/query.ts
/** Fluent, injection-safe SQL builder — the idiomatic Node-side mirror of the
 * Rust `Query` builder. Produces `{ sql, params }` for `Client.query`. */

export type Param = number | bigint | string | boolean | null;

export class Query {
  private cols: string[] = [];
  private wheres: string[] = [];
  private params: Param[] = [];
  private orderBy?: string;
  private limitN?: number;
  private offsetN?: number;

  private constructor(private readonly tableName: string) {}

  static table(name: string): Query {
    return new Query(name);
  }

  select(...columns: string[]): this {
    this.cols = columns;
    return this;
  }

  /** Add a `WHERE` predicate; one `?` per supplied param. ANDed together. */
  filter(predicate: string, ...params: Param[]): this {
    this.wheres.push(predicate);
    this.params.push(...params);
    return this;
  }

  order(column: string, direction: "ASC" | "DESC" = "ASC"): this {
    this.orderBy = `${column} ${direction}`;
    return this;
  }

  limit(n: number): this {
    this.limitN = n;
    return this;
  }

  offset(n: number): this {
    this.offsetN = n;
    return this;
  }

  build(): { sql: string; params: Param[] } {
    const cols = this.cols.length ? this.cols.join(", ") : "*";
    let sql = `SELECT ${cols} FROM ${this.tableName}`;
    if (this.wheres.length) sql += ` WHERE ${this.wheres.join(" AND ")}`;
    if (this.orderBy) sql += ` ORDER BY ${this.orderBy}`;
    if (this.limitN !== undefined) sql += ` LIMIT ${this.limitN}`;
    if (this.offsetN !== undefined) sql += ` OFFSET ${this.offsetN}`;
    return { sql, params: this.params };
  }
}
